### PR TITLE
executor: Introduce Coverage Classes

### PR DESCRIPTION
**Pull request by retrage**
## Type of PR

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue

#30 

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

For the background, see #30.
This PR proposes introducing `ShmCoverage` coverage class and the family to represent coverage stored in shared memory. All the executors that gives coverage information should inherit these coverage classes to clarify what types of coverage is supported by the executor.

## Concerns (Optional)

- [ ] Performance
- [x] Source Code Quality

This PR change adds multiple inheritance in executor classes. Multiple inheritance sometimes causes problems because it's hard to follow the class relations. I made efforts to simplify the coverage classes as much as possible, not I'm not sure this design is clear to other developers.

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

**Comment by potetisensei** [comment]
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [x] The reviewer assigned more reviewers if needed.
- [x] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [x] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

File: executor/pt_executor.cpp [hunk 1]
```diff
@@ -31,10 +31,10 @@ PTExecutor::PTExecutor(
...
                     path_to_write_input, 0, 0, record_stdout_and_err ),
-    path_shm_size ( PTExecutor::PATH_SHM_SIZE ),
-    fav_shm_size ( PTExecutor::FAV_SHM_SIZE )
+    AFLPTPathCoverage(PTExecutor::PATH_SHM_SIZE),
+    AFLPTPathFav(PTExecutor::FAV_SHM_SIZE)
 {
-    if (path_shm_size > 0 || fav_shm_size > 0) {
+    if (AFLPTPathCoverage::GetMapSize() > 0 || AFLPTPathFav::GetMapSize() > 0) {
         has_shared_memories = true;
     }
```
Comment (potetisensei) [inline executor/pt_executor.cpp]: It seems `has_shared_memories` is no longer used in `PTExecutor`. This is not so related to this PR, but probably we can remove it.

Comment (potetisensei) [reply executor/pt_executor.cpp]: Oops, sorry. PTExecutor inherits from ProxyExecutor. Never mind.

File: include/fuzzuf/coverage/shm_coverage.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    return map_size;
+  }
+
+  virtual int GetShmID(void) {
+    return shmid;
+  }
+
+  InplaceMemoryFeedback GetFeedback(std::shared_ptr<u8> lock) {
+    return InplaceMemoryFeedback(trace_bits, map_size, lock);
+  }
```
Comment (potetisensei) [inline include/fuzzuf/coverage/shm_coverage.hpp]: I think it is `ShmCoverage` that should have `lock`, not executors. If some executor wrongly passes different locks to the same instance of this class in this function, then the returned feedbacks have different locks and hence we can overwrite the shared memory although it may be referenced to in multiple locations.

File: include/fuzzuf/executor/native_linux_executor.hpp [hunk 3]
```diff
@@ -38,42 +40,24 @@
 // Responsibility (TODO):
 //  - The lifetime for the class itself and the time period of validity of member variables must match (TODO because checks have not completed)
 //      - This is to ensure the robustness
-class NativeLinuxExecutor : public Executor {
+class NativeLinuxExecutor : public Executor, public AFLEdgeCoverage, public FuzzufBBCoverage {
```
Comment (potetisensei) [inline include/fuzzuf/executor/native_linux_executor.hpp]: Do these classes really have to be inherited from?  I feel executor is not one kind of "something that creates and manages shared memory". Rather, "something that creates and manages shared memory" would be naturally a component constituting Executor, and thus composition can be more appropriate rather than Inheritance.

File: include/fuzzuf/coverage/shm_coverage.hpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/logger/logger.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+/**
+ * @class ShmCoverage
+ * @brief Base class that represents coverage stored in shared memory.
+ */
+class ShmCoverage {
```
Comment (potetisensei) [inline include/fuzzuf/coverage/shm_coverage.hpp]: I feel naming this kind of classes `FoobarCoverage` is a bit ambiguous. What this class does is setting up shared memory and maintaining it to let other classes obtain coverage information. In other words, this class does not express coverage itself, and there may be other more comprehensible names.

**Review by potetisensei — CHANGES_REQUESTED** [review]
This PR is refactoring executors, which makes much clearer the design of executors and the mechanism of obtaining coverage. 
While mostly I agree to this modification, I suggest a few changes in some little parts.

File: include/fuzzuf/coverage/shm_coverage.hpp (hunk 4 above)
Comment (potetisensei) [reply include/fuzzuf/coverage/shm_coverage.hpp]: Say, something like `ShmCoverageMaintainer`? Nope this sounds lame and dumb.
Anyways any suggestion for another name?

Comment (retrage) [reply include/fuzzuf/coverage/shm_coverage.hpp]: I agree. These classes need to have a suffix to represent their role. The candidates are: `Manager`, `Provider`, `Controller`. I prefer to use `Manager` because it is widely used, but I'm afraid that it is unclear since this word is often overused.

File: include/fuzzuf/executor/native_linux_executor.hpp (hunk 3 above)
Comment (retrage) [reply include/fuzzuf/executor/native_linux_executor.hpp]: Fixed with [8a1bff5](https://github.com/fuzzuf/fuzzuf/pull/34/commits/8a1bff54f87568299cf3cf66b8237d630c5f2804).

File: include/fuzzuf/coverage/shm_coverage.hpp (hunk 4 above)
Comment (potetisensei) [reply include/fuzzuf/coverage/shm_coverage.hpp]: Yeah I have the same concern. It would be preferable if there is a more concrete name.
How about `ShmCoverageAttacher`? I'm imagining the class to attach new shared memory to executors for adding an extra feedback. A small problem is that the class actually manages the shared memory later, not just attaching it and leaving it alone. So it may be misleading still.

File: include/fuzzuf/executor/native_linux_executor.hpp (hunk 3 above)
Comment (potetisensei) [reply include/fuzzuf/executor/native_linux_executor.hpp]: Nice.

File: include/fuzzuf/coverage/shm_coverage.hpp (hunk 4 above)
Comment (ntddk) [reply include/fuzzuf/coverage/shm_coverage.hpp]: :+1: to `ShmCoverageAttacher`.
Once I thought of `ShmCoverageDispatcher`, yet its inappropriate since Coverage itself is not a function.

Comment (retrage) [reply include/fuzzuf/coverage/shm_coverage.hpp]: Thanks for the comment. I'll add changes to use `Attacher` for the classes.

Comment (retrage) [reply include/fuzzuf/coverage/shm_coverage.hpp]: Applied with [cc2c909](https://github.com/fuzzuf/fuzzuf/pull/34/commits/cc2c90959118fee39622228c96aba8f62c43664c).

File: include/fuzzuf/coverage/shm_coverage.hpp (hunk 2 above)
Comment (retrage) [reply include/fuzzuf/coverage/shm_coverage.hpp]: I changed to use each coverage attacher has `lock`, and the executor refers to the locks in commit 
77a8053.

File: executor/proxy_executor.cpp [hunk 5]
```diff
@@ -388,7 +384,7 @@ namespace detail {
  */
 void ProxyExecutor::Run(const u8 *buf, u32 len, u32 timeout_ms) {
     // locked until std::shared_ptr<u8> lock is used in other places
-    while (lock.use_count() > 1) {
+    while (lock.use_count() > 1 || afl_edge_coverage.GetLockUseCount() > 1 || fuzzuf_bb_coverage.GetLockUseCount() > 1) {
```
Comment (potetisensei) [inline executor/proxy_executor.cpp]: Here you can use `IsFeedbackLocked`.

File: executor/native_linux_executor.cpp [hunk 6]
```diff
@@ -717,54 +729,29 @@ ExitStatusFeedback NativeLinuxExecutor::GetExitStatusFeedback() {
     return ExitStatusFeedback(last_exit_reason, last_signal);
 }
 
+bool NativeLinuxExecutor::IsFeedbackLocked() {
+    return (lock.use_count() > 1)
```
Comment (potetisensei) [inline executor/native_linux_executor.cpp:733]: Is this `NativeLinuxExecutor::lock` still used? If not, I think we can remove it.

**Review by potetisensei — CHANGES_REQUESTED** [review]
It looks almost perfect. 
I would like to suggest a few minor changes.

File: executor/native_linux_executor.cpp (hunk 6 above)
Comment (potetisensei) [reply executor/native_linux_executor.cpp:733]: The same for other executors.

Comment (retrage) [reply executor/native_linux_executor.cpp:733]: The `NativeLinuxExecutor::lock` is still used by `GetStdOut()` and `GetStdErr()`. If `stdout_buffer` and `stderr_buffer` (`std::vector<std::uint8_t>`) can have own locks, then we can remove it.

File: executor/proxy_executor.cpp (hunk 5 above)
Comment (retrage) [reply executor/proxy_executor.cpp]: Oops, sorry. Fixed with 291f46f.

File: executor/native_linux_executor.cpp (hunk 6 above)
Comment (potetisensei) [reply executor/native_linux_executor.cpp:733]: Oh sorry I forgot it. Okay let's leave it then.

**Review by potetisensei — APPROVED** [review]
Great job!

**Comment by Fadis** [comment]
run tests
